### DISCUSSION TITLE
importing Operators directly from DAGS folder in airflow

### DISCUSSION THREAD

**Discussion by madhusudhan0542** [apache/airflow#17569] [Q&A]
My intent is to create a Operator folder in dags folder (airflow/dags/operator/) and place my operators in use there. so that I can create a repository at git for dags folder and when I make changes to operator which can reflect to the repository and then the team members can also clone accordingly

airflow
 |--dags
 |      |--operators
 |      |       |--MsSqlOperator.py
 |      |--dag_file.py
 |      |--dag_file1.py
 |      |--dag_file2.py
 |

here while I was trying to import Class from the MsSqlOperator.py I was unable to import, were it states MsSqlOperator is undefined because airflow can't point to dags folder while importing , How can I import my operator file which is in dags/operators/  folder

thank you in advance!

**Comment by potiuk** [comment]
You miss empty `__init__.py` in your `operators` folder. It will turn `operators` into python package and you will be able to use:

`from operators.MsSqlOperator import ...`

in your dags.

Few points though:
* don't use capital letters for your module name (python convention is lower_letters_with_underscore)
* add .airflowignore file in your `dags` folder and add `operators/.*` there. This will tell airflow to skip the whole `operators` folder when scanning for dags.

**Reply by madhusudhan0542** [reply]
ok, now I had created the .airflowignore in dags  and underscore on both sides __init__.py file in my operators and made them to assume as python packages,
and tried with      "from operators.MsSqlOperator import ms_sql"  but still it says     'NameError: name 'ms_sql' is not defined'

here my doubt is, how will we indicate airflow to import from operators folder 'dags/operators/' as "from operators.sftp_to_s3_operator import ..." were it is initially importing from 'airflow/operators/...'  as for an example  "from airflow.contrib.operators.sftp_to_s3_operator import SFTPToS3Operator"

**Reply by potiuk** [reply]
The "dags' are added to PYTHONPATH (in Airlfow 2 at least, I am not sure how it was in 1.10 - so my advise is not applicable to it.).. This is basic Python behaviour - it will import from anywhere where PYTHONPATH point to.

BTW if you are in Airflow 1.10 still - migrate to Airflow 2  - Airflow 1.10 is end-of-life as of June 17 and will not get any more fixes and support.  You can read more how it is in Airllow 2 here: https://airflow.apache.org/docs/apache-airflow/stable/modules_management.html  - and you can find all your answers there (providing you migrate).

if you do not find ms_sql in your module, maybe it's not defined there? I cannot say - I do not see you MsqSql.py (which - again, let me repeat that - should not be named with capital case but with lower_case_with_underscores. Using case might cause various problems if you use filesystem that is case-insensitve so naming your module name with CamelCase is really bad idea.

My advise: use different name than "operators" (my_custom_operators or smth). to not confuse it with anything else - this way you will be sure you are importing the right thing and not other "operators" accidentally.

**Reply by jedcunningham** [reply]
I just want to add a +1 to what @potiuk said: Airflow just follows the normal python import process - "dags" is automatically added to sys.path and at that point it's just a normal import.

Are you still having issues?

**Reply by madhusudhan0542** [reply]
No I don't have any issues now, The same mentioned way we can import them by just giving  "from operators.MsSqlOperator import ms_sql" , in addition all the permissions to the file mentioned should be given(reason why I can't import earlier), Now I can Import them and thank you very much for the assistance
